homeassistant/components/yolink/__init__.py

```diff
@@ -50,58 +71,20 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     auth_mgr = api.ConfigEntryAuth(
         hass, aiohttp_client.async_get_clientsession(hass), session
     )
-
-    yolink_http_client = YoLinkClient(auth_mgr)
-    yolink_mqtt_client = MqttClient(auth_mgr)
-
-    def on_message_callback(message: tuple[str, BRDP]) -> None:
-        data = message[1]
-        device_id = message[0]
-        if data.event is None:
-            return
-        event_param = data.event.split(".")
-        event_type = event_param[len(event_param) - 1]
-        if event_type not in (
-            "Report",
-            "Alert",
-            "StatusChange",
-            "getState",
-        ):
-            return
-        resolved_state = data.data
-        if resolved_state is None:
-            return
-        entry_data = hass.data[DOMAIN].get(entry.entry_id)
-        if entry_data is None:
-            return
-        device_coordinators = entry_data.get(ATTR_COORDINATORS)
-        if device_coordinators is None:
-            return
-        device_coordinator = device_coordinators.get(device_id)
-        if device_coordinator is None:
-            return
-        device_coordinator.async_set_updated_data(resolved_state)
-
+    yolink_home = YoLinkHome()
     try:
         async with async_timeout.timeout(10):
-            device_response = await yolink_http_client.get_auth_devices()
-            home_info = await yolink_http_client.get_general_info()
-            await yolink_mqtt_client.init_home_connection(
-                home_info.data["id"], on_message_callback
+            await yolink_home.async_setup(
+                auth_mgr, YoLinkHomeMessageListener(hass, entry)
             )
     except YoLinkAuthFailError as yl_auth_err:
         raise ConfigEntryAuthFailed from yl_auth_err
     except (YoLinkClientError, asyncio.TimeoutError) as err:
         raise ConfigEntryNotReady from err
 
-    hass.data[DOMAIN][entry.entry_id] = {
-        ATTR_CLIENT: yolink_http_client,
-        ATTR_MQTT_CLIENT: yolink_mqtt_client,
-    }
-    auth_devices = device_response.data[ATTR_DEVICE]
+    hass.data[DOMAIN][entry.entry_id] = {ATTR_HOME_INSTANCE: yolink_home}
```

Comment:
I recommend using a dataclass to store multiple items in hass.data.

https://docs.python.org/3/library/dataclasses.html